py/desiutil/redirect.py

```diff
@@ -0,0 +1,247 @@
+            hformat = None
+            while len(logger.handlers) > 0:
+                h = logger.handlers[0]
+                if hformat is None:
+                    hformat = h.formatter._fmt
+                logger.removeHandler(h)
+            # Add the current stdout.
+            ch = logging.StreamHandler(sys.stdout)
+            formatter = logging.Formatter(hformat, datefmt='%Y-%m-%dT%H:%M:%S')
+            ch.setFormatter(formatter)
+            logger.addHandler(ch)
+
+    def _open_redirect(filename):
+        # Open python file, which creates low-level POSIX file
+        # descriptor.
+        file_handle = open(filename, "w")
+
+        # Redirect stdout/stderr to this new file descriptor.
+        _redirect(out_to=file_handle.fileno(), err_to=file_handle.fileno())
+        return file_handle
+
+    def _close_redirect(handle):
+        # Close python file handle, which will mark POSIX file descriptor for
+        # garbage collection.  That is fine since we are about to overwrite those.
+        if handle is not None:
+            handle.close()
+
+        # Flush python handles for good measure
+        sys.stdout.flush()
+        sys.stderr.flush()
+
+        try:
+            # Restore old stdout and stderr
+            _redirect(out_to=saved_fd_out, err_to=saved_fd_err)
+        except:
+            pass
+
+    # Redirect both stdout and stderr to the same file
+
+    if to is None:
+        to = "/dev/null"
+
+    if rank == 0:
+        log = get_logger()
+        log.info("Begin log redirection to {} at {}".format(to, time.asctime()))
+
+    # Try to open the redirected file.
+
+    pto = to
+    if to != "/dev/null" and comm is not None:
+        pto = "{}_{}".format(to, rank)
+
+    fail_open = 0
+    file = None
+    try:
+        file = _open_redirect(pto)
+    except:
+        log = get_logger()
+        log.error(
+            "Failed to open redirection file {} at {}".format(pto, time.asctime())
+        )
+        fail_open = 1
+
+    if comm is not None:
+        fail_open = comm.allreduce(fail_open, op=MPI.SUM)
+
+    if fail_open > 0:
+        # Something went wrong on one or more processes, try to recover and exit
+        if rank == 0:
+            log = get_logger()
+            log.error(
+                "Failed to start redirect to {} at {}".format(to, time.asctime())
+            )
+
+        _close_redirect(file)
+
+        # All processes raise an exception for the calling code to handle
+        msg = "Failed to start output redirect to {}".format(to)
+        raise RuntimeError(msg)
+
+    # Output should now be redirected.  Run the code.
+
+    fail_run = 0
+    try:
+        yield # Allow code to be run with the redirected output
+    except:
+        # We have an unhandled exception.  Print a stack trace to the log.
+        exc_type, exc_value, exc_traceback = sys.exc_info()
+        lines = traceback.format_exception(exc_type, exc_value, exc_traceback)
+        print("".join(lines), flush=True)
```

Comment:
This bare `print()` is safe?

Reply:
This occurs while the stdout / stderr is being redirected to a file, and the wrapped code block just raised an exception.  So the stack trace will end up in the redirected file.  I think it is fine...